ntfs3g: use external FUSE, set meta.mainProgram

howtonotwin opened the pull request:
On Linux, by default, the ntfs-3g binaries attempt to perform `mount()` by themselves (through the included libfuse-lite). Thus, the binaries will not function when run unprivileged. Usually, they are installed setuid root, but Nix clears the setuid bit on everything in the store, so ntfs-3g as built by us was unusable for unprivileged users on Linux.

In order to let unprivileged users use ntfs-3g, configure the build to use the "external" FUSE implementation `pkgs.fuse`, which arranges for its own binaries to be setuid root.

Also set `meta.mainProgram`. The net effect is that an unprivileged user can `nix run nixpkgs#ntfs3g` and it will mostly just work. (`-o no_def_opts` also needs to be passed to avoid ntfs-3g setting `-o allow_other`.)

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

SigmaSquadron reviewed (dismissed):
Welcome to Nixpkgs! Thanks for improving `ntfs3g`.

Could the `meta.mainProgram` commit be split into a separate one? The changes don't depend on each other.

howtonotwin commented on pkgs/by-name/nt/ntfs3g/package.nix:
```diff
@@ -68,6 +72,7 @@ stdenv.mkDerivation rec {
     "--enable-extras"
     "--with-mount-helper=${mount}/bin/mount"
     "--with-umount-helper=${mount}/bin/umount"
+    "--with-fuse=external"
```
`--with-fuse` only accepts the values `internal` and `external`. `external` is the default on all platforms except Linux (and Solaris) already.

When `--with-fuse=external`, ntfs-3g uses pkg-config to look for libfuse. That should find the one in the store as long as `fuse` is in `buildInputs`, right?

howtonotwin committed:
ntfs3g: set meta.mainProgram

howtonotwin committed:
ntfs3g: use external FUSE

On Linux, by default, the ntfs-3g binaries attempt to perform mount() by
themselves (through the included libfuse-lite library). Thus, for an
unprivileged user to use ntfs-3g, the binaries must be setuid root. But
Nix clears the setuid bit on everything in the store, so ntfs-3g is
unusable except by root on Linux.

In order to let unprivileged users use ntfs-3g, configure the build to
use the "external" FUSE implementation pkgs.fuse, which arranges for its
own binaries to be setuid root.

An unprivileged user can now nix run nixpkgs#ntfs3g and it should mostly
"just work".

buildInputs is simplified because pkgs.fuse is already a conditional
that picks between macfuse-stubs (on Darwin) and fuse_2 (on Linux).

SigmaSquadron commented:
> @SigmaSquadron your own review is blocking, auto-merge won't do it until that's dismissed / approved.

Yeah, that's my bad. I was running a Nixpkgs review that would've auto-approved the PR when done and merged it, but it looks like it didn't work!
